compiler/rustc_builtin_macros/src/offset_of.rs

```diff
@@ -0,0 +1,83 @@
+use rustc_ast as ast;
+use rustc_ast::ptr::P;
+use rustc_ast::token;
+use rustc_ast::tokenstream::TokenStream;
+use rustc_ast_pretty::pprust;
+use rustc_errors::PResult;
+use rustc_expand::base::{self, *};
+use rustc_parse::parser::Parser;
+use rustc_span::{symbol::Ident, Span};
+
+fn parse_field<'a>(p: &mut Parser<'a>) -> PResult<'a, Ident> {
+    let token = p.token.uninterpolate();
+    let field = match token.kind {
+        token::Ident(name, _) => Ok(Ident::new(name, token.span)),
+        token::Literal(token::Lit { kind: token::Integer, symbol, suffix: None }) => {
+            Ok(Ident::new(symbol, token.span))
+        }
+        token::Eof => Err(p.struct_span_err(p.token.span, "offset_of! takes 2 arguments`")),
+        _ => {
+            let actual = pprust::token_to_string(&p.token);
+            Err(p.struct_span_err(p.token.span, &format!("unexpected token: `{actual}`")))
+        }
+    }?;
+
+    p.bump();
+
+    Ok(field)
+}
+
+fn parse_args<'a>(
+    cx: &mut ExtCtxt<'a>,
+    sp: Span,
+    tts: TokenStream,
+) -> PResult<'a, (P<ast::Ty>, Vec<Ident>)> {
+    let mut p = cx.new_parser_from_tts(tts);
+
+    let container = p.parse_ty()?;
+
+    p.expect(&token::Comma)?;
+
+    let mut fields = Vec::new();
+
+    loop {
+        let field = parse_field(&mut p)?;
+
+        fields.push(field);
+        if p.eat(&token::Dot) {
+            continue;
+        }
+
+        if p.token != token::Eof && p.token != token::Comma {
```

Comment:
this appears to allow things like `offset_of!(A, b.c, junk :P)` since it doesn't check beyond the comma after the fields -- though i could be misunderstanding the parser